rewrite-java-test/src/test/java/org/openrewrite/java/cleanup/ReplaceRedundantFormatWithPrintfTest.java

```diff
@@ -0,0 +1,293 @@
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    void appendsNewlineForPrintlnWithTextBlockFormatString() {
+        rewriteRun(
+          java(
+            """
+              class Test {
+                  void test(String arg) {
+                      System.out.println(String.format(\"\"\"
+                      hello %s\"\"\", arg));
+                  }
+              }
+              """,
+            """
+              class Test {
+                  void test(String arg) {
+                      System.out.printf(\"\"\"
+                      hello %s%n\"\"\", arg);
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    /**
+     * Tests that the code generation used for JavaTemplate behaves correctly given a parameterized type.
+     */
+    void doesNotFailWhenArgHasParameterizedType() {
+        rewriteRun(
+          java(
+            """
+              class Test {
+                  void test() {
+                      java.util.List<Integer> ints = new java.util.ArrayList<>();
+                      System.out.print(String.format("hello %s", ints));
+                  }
+              }
+              """,
+            """
+              class Test {
+                  void test() {
+                      java.util.List<Integer> ints = new java.util.ArrayList<>();
+                      System.out.printf("hello %s", ints);
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    /**
+     * Tests that the code generation used for JavaTemplate behaves correctly given a primitive type.
+     */
+    void doesNotFailWhenArgHasPrimitiveType() {
+        rewriteRun(
+          java(
+            """
+              class Test {
+                  void test() {
+                      int value = 42;
+                      System.out.print(String.format("hello %i", value));
+                  }
+              }
+              """,
+            """
+              class Test {
+                  void test() {
+                      int value = 42;
+                      System.out.printf("hello %i", value);
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    /**
+     * Tests that the code generation used for JavaTemplate behaves correctly given a template parameter.
+     */
+    void doesNotFailWhenArgHasTemplateParameter() {
+        rewriteRun(
+          java(
+            """
+              class Test<T> {
+                  void test(T arg) {
+                      System.out.print(String.format("hello %s", arg));
+                  }
+              }
+              """,
+            """
+              class Test<T> {
+                  void test(T arg) {
+                      System.out.printf("hello %s", arg);
+                  }
+              }
+              """
+          )
+        );
+    }
+}
```

Comment:
not sure if these tests are gratuitous; I was nervous about how I built the code for the `JavaTemplate`